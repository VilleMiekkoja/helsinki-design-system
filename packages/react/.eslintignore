**/*.d.ts
webpack.config.js
.templates
lib/
