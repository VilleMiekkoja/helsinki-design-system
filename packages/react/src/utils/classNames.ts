import classnames from 'classnames/dedupe';

export default classnames;
