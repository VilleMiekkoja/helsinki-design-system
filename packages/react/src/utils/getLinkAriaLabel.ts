import React from 'react';

export default (
  ariaLabel: string,
  label: React.ReactNode,
  openInNewTabAriaLabel?: string,
  openInExternalDomainAriaLabel?: string,
  external?: boolean,
  openInNewTab?: boolean,
) => {
  let childrenText = ariaLabel || label;
  const newTabText = openInNewTab ? openInNewTabAriaLabel || 'Avautuu uudessa välilehdessä.' : '';
  const externalText = external ? openInExternalDomainAriaLabel || 'Siirtyy toiseen sivustoon.' : '';

  if (childrenText && childrenText.slice(-1) !== '.') {
    childrenText = `${childrenText}.`;
  }

  return [childrenText, newTabText, externalText].filter((text) => text).join(' ');
};
