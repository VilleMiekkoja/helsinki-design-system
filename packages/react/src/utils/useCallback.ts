import { EventHandler, useCallback } from 'react';

export const useCallbackIfDefined = (callback) => {
  const handler = (event) => {
    if (callback) callback(event);
  };

  return useCallback(handler, [callback]);
};

export const useEnterOrSpacePressCallback = (callback) => {
  const handler = (event) => {
    if (event.key === 'Enter' || event.key === ' ') callback(event);
  };

  return useCallbackIfDefined(handler);
};

export const withDefaultPrevented = (callback: EventHandler<never>) =>
  function patchedCallback(event) {
    event.preventDefault();
    return callback.call(callback, event);
  };

export const withPropagationStopping = (callback: EventHandler<never>) =>
  function patchedCallback(event) {
    event.stopPropagation();
    return callback.call(callback, event);
  };
