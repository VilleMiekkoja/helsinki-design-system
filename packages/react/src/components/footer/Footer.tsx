import React from 'react';

// import core base styles
import 'hds-core';
import styles from './Footer.module.scss';
import { Koros, KorosType } from '../koros';
import classNames from '../../utils/classNames';
import { FooterNavigation } from './footerNavigation/FooterNavigation';
import { FooterNavigationGroup } from './footerNavigationGroup/FooterNavigationGroup';
import { FooterNavigationHeading } from './footerNavigationHeading/FooterNavigationHeading';
import { FooterNavigationLink } from './footerNavigationLink/FooterNavigationLink';
import { FooterUtilities } from './footerUtilities/FooterUtilities';
import { FooterUtilityGroup } from './footerUtilityGroup/FooterUtilityGroup';
import { FooterSoMe } from './footerSoMe/FooterSoMe';
import { FooterBase } from './footerBase/FooterBase';
import { FooterCustom } from './footerCustom/FooterCustom';
import { FooterTheme } from './Footer.interface';
import { getComponentFromChildren } from '../../utils/getChildren';
import { useTheme } from '../../hooks/useTheme';

export type FooterProps = React.PropsWithChildren<{
  /**
   * Additional class names to apply to the footer
   */
  className?: string;
  /**
   * Props that will be passed to the native `<footer>` element
   */
  footerProps?: React.ComponentPropsWithoutRef<'footer'>;
  /**
   * Koros type to use in the footer
   */
  korosType?: KorosType;
  /**
   * Defines the footer theme
   */
  theme?: FooterTheme;
  /**
   * The title of the service shown on top of the the footer
   */
  title?: React.ReactNode;
}>;

export const Footer = ({
  children,
  className,
  footerProps,
  korosType = 'basic',
  theme = 'light',
  title,
}: FooterProps) => {
  // custom theme class that is applied to the root element
  const customThemeClass = useTheme<FooterTheme>(styles.footer, theme);
  // filter out navigation from other children so that they can be rendered separately
  const [navigation, childrenWithoutNavigation] = getComponentFromChildren(children, 'FooterNavigation');

  return (
    <footer
      {...footerProps}
      className={classNames(
        styles.footer,
        styles[`koros-${korosType}`],
        typeof theme === 'string' && styles[`theme-${theme}`],
        customThemeClass,
        className,
      )}
    >
      <Koros className={classNames(styles.koros, styles[korosType])} type={korosType} />
      <div className={styles.footerContent}>
        <section className={classNames(styles.navigationContainer, styles.centeredNavigation)}>
          <div className={styles.titleWrapper}>{title && <h2 className={styles.title}>{title}</h2>}</div>
          {navigation}
        </section>
        {childrenWithoutNavigation}
      </div>
    </footer>
  );
};

Footer.Navigation = FooterNavigation;
Footer.NavigationGroup = FooterNavigationGroup;
Footer.NavigationHeading = FooterNavigationHeading;
Footer.NavigationLink = FooterNavigationLink;
Footer.Utilities = FooterUtilities;
Footer.UtilityGroup = FooterUtilityGroup;
Footer.SoMe = FooterSoMe;
Footer.Base = FooterBase;
Footer.Custom = FooterCustom;
