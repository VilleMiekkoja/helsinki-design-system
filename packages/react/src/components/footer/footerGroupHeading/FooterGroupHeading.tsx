import React from 'react';

// import core base styles
import 'hds-core';
import styles from './FooterGroupHeading.module.scss';
import { MergeElementProps } from '../../../common/types';
import classNames from '../../../utils/classNames';
import { FooterVariant } from '../Footer.interface';

type ItemProps<Element> = React.PropsWithChildren<{
  /**
   * Element type
   */
  as?: Element;
  /**
   * Hypertext Reference of the heading link.
   */
  href?: string;
  /**
   * Id for the heading element.
   */
  id?: string;
  /**
   * The label for the heading.
   */
  label?: string | React.ReactNode;
  /**
   * Internal variant to change styles based on context.
   */
  variant?: FooterVariant.Navigation | FooterVariant.Utility;
}>;

export type FooterGroupHeadingProps<Element extends React.ElementType = 'a'> = MergeElementProps<
  Element,
  ItemProps<Element>
>;

export const FooterGroupHeading = <T extends React.ElementType = 'a'>({
  as,
  children,
  className,
  href,
  label,
  variant,
  ...rest
}: FooterGroupHeadingProps<T>) => {
  const Item: React.ElementType = as || (href ? 'a' : 'span');
  return (
    <Item className={classNames(styles.heading, variant && styles[variant], className)} href={href} {...rest}>
      {label && label}
      {children}
    </Item>
  );
};
FooterGroupHeading.componentName = 'FooterGroupHeading';
