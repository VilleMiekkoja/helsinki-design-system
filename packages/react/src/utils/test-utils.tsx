import React, { PropsWithChildren } from 'react';

import { Navigation } from '../components/navigation';
import { Footer } from '../components/footer';
import { Header } from '../components/header';

type WrapperProps = PropsWithChildren<Record<string, unknown>>;

export const NavigationWrapper = ({ children }: WrapperProps) => (
  <Navigation menuToggleAriaLabel="menu" skipTo="#content" skipToContentLabel="Skip to content" title="Foo">
    {children}
  </Navigation>
);

export const FooterWrapper = ({ children }: WrapperProps) => <Footer title="Bar">{children}</Footer>;

export const HeaderWrapper = ({ children }: WrapperProps) => <Header>{children}</Header>;

export const HeaderNavigationMenuWrapper = ({ children }: WrapperProps) => (
  <Header>
    <Header.NavigationMenu>{children}</Header.NavigationMenu>
  </Header>
);
