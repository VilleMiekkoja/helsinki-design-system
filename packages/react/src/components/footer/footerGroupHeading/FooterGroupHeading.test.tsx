import React from 'react';
import { render } from '@testing-library/react';
import { axe } from 'jest-axe';

import { FooterGroupHeading } from './FooterGroupHeading';
import { FooterWrapper } from '../../../utils/test-utils';
import { FooterVariant } from '../Footer.interface';

describe('<Footer.FooterGroupHeading /> spec', () => {
  it('renders the component', () => {
    const { asFragment } = render(<FooterGroupHeading label="Test" />, { wrapper: FooterWrapper });
    expect(asFragment()).toMatchSnapshot();
  });

  it('should not have basic accessibility issues', async () => {
    const { container } = render(<FooterGroupHeading label="Test" />, { wrapper: FooterWrapper });
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });

  it('renders the utility variant', () => {
    const { asFragment } = render(<FooterGroupHeading label="Test" variant={FooterVariant.Utility} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders the navigation variant', () => {
    const { asFragment } = render(<FooterGroupHeading label="Test" variant={FooterVariant.Navigation} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
