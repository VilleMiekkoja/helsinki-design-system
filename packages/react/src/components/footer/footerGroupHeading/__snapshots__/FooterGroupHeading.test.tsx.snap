// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Footer.FooterGroupHeading /> spec renders the component 1`] = `
<DocumentFragment>
  <footer
    class="footer koros-basic theme-light"
  >
    <div
      class="koros basic koros basic"
    >
      <svg
        aria-hidden="true"
        height="85"
        width="100%"
        xmlns="http://www.w3.org/2000/svg"
      >
        <defs>
          <pattern
            height="85"
            id="koros_basic-1"
            patternUnits="userSpaceOnUse"
            width="106"
            x="0"
            y="0"
          >
            <path
              d="M0,800h20V0c-4.9,0-5,2.6-9.9,2.6S5,0,0,0V800z"
              transform="scale(5.3)"
            />
          </pattern>
        </defs>
        <rect
          fill="url(#koros_basic-1)"
          height="85"
          width="100%"
        />
      </svg>
    </div>
    <div
      class="footerContent"
    >
      <div
        class="footerSections"
      >
        <h2
          class="title"
        >
          Bar
        </h2>
        <span
          class="heading"
        >
          Test
        </span>
      </div>
    </div>
  </footer>
</DocumentFragment>
`;

exports[`<Footer.FooterGroupHeading /> spec renders the navigation variant 1`] = `
<DocumentFragment>
  <span
    class="heading navigation"
  >
    Test
  </span>
</DocumentFragment>
`;

exports[`<Footer.FooterGroupHeading /> spec renders the utility variant 1`] = `
<DocumentFragment>
  <span
    class="heading utility"
  >
    Test
  </span>
</DocumentFragment>
`;
